index.d.ts

```diff
@@ -0,0 +1,11 @@
+interface DotPropImmutable {
+  get<I, T>(source: I, path: number | string | (string | number)[], defaultValue?: any): T;
+  set<I, T>(source: I, path: number | string | (string | number)[], value: any): T;
```

Comment:
Why `I != T`, shouldn't you get back the same type as you put in ? Also value could be type `Partial(T)` ??

Delete might return a partial also since type is mostly the same just missing a key??

Just some thoughts :)